grpc/src/main/java/io/stargate/grpc/codec/cql/UdtCodec.java

```diff
@@ -0,0 +1,84 @@
+package io.stargate.grpc.codec.cql;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.stargate.db.schema.Column;
+import io.stargate.db.schema.Column.ColumnType;
+import io.stargate.db.schema.UserDefinedType;
+import io.stargate.proto.QueryOuterClass;
+import io.stargate.proto.QueryOuterClass.UdtValue;
+import io.stargate.proto.QueryOuterClass.Value;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class UdtCodec extends CompositeCodec {
+
+  @Override
+  public ByteBuffer encode(@NonNull QueryOuterClass.Value value, @NonNull Column.ColumnType type) {
+    if (!value.hasUdt()) {
+      throw new IllegalArgumentException("Expected user-defined type");
+    }
+    assert type.isUserDefined();
+
+    UdtValue udtValue = value.getUdt();
+    Map<String, Value> udtFieldsValues = udtValue.getFieldsMap();
+
+    Map<String, Column> fields = ((UserDefinedType) type).columnMap();
+    for (String name : udtFieldsValues.keySet()) {
+      if (!fields.containsKey(name)) {
+        throw new IllegalArgumentException(
+            String.format("User-defined type doesn't contain a field named '%s'", name));
+      }
+    }
+
+    int fieldsCount = fields.size();
+    ByteBuffer[] encodedFields = new ByteBuffer[fieldsCount];
+
+    int i = 0;
+    int toAllocate = 0;
+    for (Column field : fields.values()) {
+      ByteBuffer encodedField = null;
+      Value fieldValue = udtFieldsValues.get(field.name());
+      if (fieldValue != null) {
+        Column.ColumnType fieldType = Objects.requireNonNull(field.type());
+        ValueCodec fieldCodec = ValueCodecs.get(fieldType.rawType());
+        encodedField = ValueCodec.encodeValue(fieldCodec, fieldValue, fieldType);
+      }
+      toAllocate += 4 + (encodedField == null ? 0 : encodedField.remaining());
+      encodedFields[i++] = encodedField;
+    }
+
+    return encodeValues(encodedFields, ByteBuffer.allocate(toAllocate));
+  }
+
+  @Override
+  public Value decode(@NonNull ByteBuffer bytes, @NonNull Column.ColumnType type) {
+    UdtValue.Builder builder = UdtValue.newBuilder();
+
+    assert type.isUserDefined();
+
+    try {
+      ByteBuffer input = bytes.duplicate();
+      List<Column> udtFields = ((UserDefinedType) type).columns();
+      int fieldCount = udtFields.size();
+      int i = 0;
+      while (input.hasRemaining()) {
+        if (i > fieldCount) {
+          throw new IllegalArgumentException(
+              String.format("Too many fields in encoded UDT, expected %d", fieldCount));
+        }
+        String fieldName = udtFields.get(i).name();
+        ColumnType fieldType = Objects.requireNonNull(udtFields.get(i).type());
+        ValueCodec fieldCodec = ValueCodecs.get(fieldType.rawType());
+        builder.putFields(fieldName, decodeValue(input, fieldCodec, fieldType));
```

Comment:
As it's currently implemented. If you have a UDT as defined: `CREATE TYPE udt (f1 int, f2 text, f3 uuid)`. If you insert a subset of the fields (which is valid) the result will return `null`s for the values that weren't provided e.g.

Insert: `Values.of(ImmutableMap.of("f2", Values.of("abc")))`
Result: `Values.of(ImmutableMap.of("f1", Values.NULL, "f2", Values.of("abc"), "f3", Values.NULL))`

There is no way to distinguish between not provided and `null` provided.

My question: Should we remove `null` values from the result? So that if we would get back `Values.of(ImmutableMap.of("f2", Values.of("abc")))` in the example above when the other fields (f1 and f3) are either `null` or not provided?

Reply:
Can we keep track of _provided_ property names and remove `null` values only for not provided properties?

Reply:
If that is not possible, I think it would still make sense to remove `null` values from the response for the same of avoiding ambiguity with respect to what is stored in the database.

Reply:
> Can we keep track of provided property names and remove null values only for not provided properties?

No. Unfortunately we *can't* track that. The storage engine just stores `null` for both actual `null` and not provided.

Reply:
I do like the idea of removing `null` values from the result, but this is not what's stored in the DB.

For reference this is what's returned from persistence:

```
token@cqlsh:testks> CREATE TABLE udt_test (key text PRIMARY KEY, value udt);
token@cqlsh:testks> INSERT INTO udt_test (key, value) VALUES ( 'k1', { f1: 1 });
token@cqlsh:testks> SELECT * FROM udt_test ;

 key | value
-----+-------------------
  k1 | {f1: 1, f2: null}

(1 rows)
token@cqlsh:testks> INSERT INTO udt_test (key, value) VALUES ( 'k1', { f2: 'v1' });
token@cqlsh:testks> SELECT * FROM udt_test ;

 key | value
-----+----------------------
  k1 | {f1: null, f2: 'v1'}

(1 rows)
token@cqlsh:testks> INSERT INTO udt_test (key, value) VALUES ( 'k1', { f1: null, f2: null });
token@cqlsh:testks> SELECT * FROM udt_test ;

 key | value
-----+-------
  k1 |  null

(1 rows)
token@cqlsh:testks> INSERT INTO udt_test (key, value) VALUES ( 'k1', { f1: 1, f2: null });
token@cqlsh:testks> SELECT * FROM udt_test ;

 key | value
-----+-------------------
  k1 | {f1: 1, f2: null}

(1 rows)
```

Reply:
Sorry, I'm bit lost here. `cqlsh` probably does not show the difference, but can we tell `UNSET` from `NULL` in our code?